api/v1alpha1/osbuildconfig_types.go

```diff
@@ -0,0 +1,133 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// OSBuildConfigSpec defines the desired state of OSBuildConfig
+type OSBuildConfigSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Details defines what to build
+	Details BuildDetails `json:"build_details"`
+	// Triggers defines when to build
+	Triggers BuildTriggers `json:"build_triggers"`
+}
+
+// BuildDetails includes all the information needed to build the image
+type BuildDetails struct {
+	// Distribution is the name of the O/S distribution
+	Distribution string `json:"distribution"`
+	// Customizations defines the changes to be applied on top of the base image (optional)
+	Customizations *Customizations `json:"customizations,omitempty"`
+	// TargetImage defines the requested output image
+	TargetImage TargetImage `json:"target_image"`
+}
+
+// Customizations defines the changes to be applied on top of the base image
+type Customizations struct {
+	// Packages is a list of RPM packages to install (optional)
+	Packages []string `json:"packages,omitempty"`
+	// Users is the list of Users to add to the image (optional)
+	Users []User `json:"users,omitempty"`
+	// Services defines the services to enable or disable (optional)
+	Services *Services `json:"services,omitempty"`
+}
+
+// User defines a single user to be configured
+type User struct {
+	// Name is the username for the new user
+	Name string `json:"name"`
+	// Groups is the groups to add the user to (optional)
+	Groups []string `json:"groups,omitempty"`
+	// Key is the user's SSH public key (optional)
+	Key *string `json:"key,omitempty"`
+}
+
+type Services struct {
+	// Enabled is the list of services to enabled (optional)
+	Enabled []string `json:"enabled,omitempty"`
+	// Disabled is the list of services to disabled (optional)
+	Disabled []string `json:"disabled,omitempty"`
+}
+
+type TargetImage struct {
+	// Architecture defines target architecture of the image
+	Architecture string `json:"architecture"`
+	// ImageType defines the target image type
+	TargetImageType TargetImageType `json:"target_image_type"`
+	// OSTree is the OSTree configuration of the build (optional)
+	OSTree *OSTreeConfig `json:"ostree,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=edge-container;edge-installer
+type TargetImageType string
+
+// OSTreeConfig defines the OSTree ref details
+type OSTreeConfig struct {
+	// Url is the URL of the target build
+	Url *string `json:"url"`
+	// Ref is the ref of the target build
+	Ref *string `json:"ref"`
+	// Parent is the ref of the parent of target build (Optional)
+	Parent *string `json:"parent"`
+}
+
+type BuildTriggers struct {
+	// ConfigChange if True trigger a new build upon any change in this BuildConfig CR (optional)
```

Comment:
No, changes in the CR will trigger the creation of a new OSBuild CR which in turn will trigger a new image build operation

Reply:
imho it would be best to make it immutable, once the CR is created it should not allow modifications: 1 CR should match 1 generated results. Otherwise we end up with a lot of logic to handle updates in the CR reconciliation.

Reply:
I see, thanks for the clarification.